sympy/core/basic.py

```diff
@@ -1813,6 +1822,20 @@ def _exec_constructor_postprocessors(cls, obj):
 
         return obj
 
+def recur_free_in(obj, syms, bs):
+    from sympy import Tuple, Integer
+    if obj in syms and obj not in bs:
+        yield obj
+    else:
+        for arg in reversed(obj.args[1:]):
```

Comment:
Given that this is a method on Basic it should not presume that the args have the structure of the args for Integral.

Reply:
@oscarbenjamin, Can you please provide some examples of such classes? 
It will be helpful to optimize the method for general `Basic`.

Some classes I can think of are:
```
Function
Derivative
Poly
Integral
Sum
Product
```
I am not sure, but do classes like the below also need `has_free`?
```
matrix
vector
```
Thanks.